src/printer-markdown.js

```diff
@@ -0,0 +1,461 @@
+"use strict";
+
+const docBuilders = require("./doc-builders");
+const concat = docBuilders.concat;
+const join = docBuilders.join;
+const line = docBuilders.line;
+const hardline = docBuilders.hardline;
+const fill = docBuilders.fill;
+const align = docBuilders.align;
+const docPrinter = require("./doc-printer");
+const printDocToString = docPrinter.printDocToString;
+const escapeStringRegexp = require("escape-string-regexp");
+
+const SINGLE_LINE_NODE_TYPES = ["heading", "tableCell", "footnoteDefinition"];
+
+const SIBLING_NODE_TYPES = ["listItem", "definition", "footnoteDefinition"];
+
+const INLINE_NODE_TYPES = [
+  "inlineCode",
+  "emphasis",
+  "strong",
+  "delete",
+  "link",
+  "linkReference",
+  "image",
+  "imageReference",
+  "footnote",
+  "footnoteReference",
+  "sentence",
+  "whitespace",
+  "word"
+];
+
+const INLINE_NODE_WRAPPER_TYPES = INLINE_NODE_TYPES.concat([
+  "tableCell",
+  "paragraph"
+]);
+
+function genericPrint(path, options, print) {
+  const node = path.getValue();
+
+  switch (node.type) {
+    case "root":
+      return normalizeDoc(
+        concat([printChildren(path, options, print), hardline])
+      );
+    case "paragraph":
+      return printChildren(path, options, print, {
+        postprocessor: fill
+      });
+    case "sentence":
+      return printChildren(path, options, print);
+    case "word":
+      return escapeHtmlEntities(escapeString(node.value, ["_", "*", "~~"]));
+    case "whitespace": {
+      return hasParentType(path, SINGLE_LINE_NODE_TYPES) ? " " : line;
+    }
+    case "emphasis":
+      return concat(["*", printChildren(path, options, print), "*"]);
+    case "strong":
+      return concat(["**", printChildren(path, options, print), "**"]);
+    case "delete":
+      return concat(["~~", printChildren(path, options, print), "~~"]);
+    case "inlineCode":
+      return concat(["`", node.value, "`"]);
+    case "link":
+      return concat([
+        "[",
+        printChildren(path, options, print),
+        "](",
+        node.url,
+        node.title ? ` "${node.title}"` : "",
+        ")"
+      ]);
+    case "image":
+      return concat([
+        "![",
+        node.alt || "",
+        "](",
+        node.url,
+        node.title ? ` "${node.title}"` : "",
+        ")"
+      ]);
+    case "blockquote":
+      return concat(["> ", align("> ", printChildren(path, options, print))]);
```

Comment:
Couldn't this be something like 

```js
concat(["> ", join([hardline, "> "],  printChildren(path, options, print)));
```

Reply:
Isn't the `join` for concat docs with specified separator? The `>` should be printed on every line break (just like `indent` but print `>`) and the `blockquote` can contains any kind of nodes.

My first attempt is to print `ifBreak("> ")` on every whitespace, but caused a lot of weird results, so I switched to use `align`.